File: Go/Lesson_100_palindrome_partitioning.go
//go:build ignore

// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 100 -- Palindrome Partitioning
// Category   : Backtracking
// Difficulty : Medium
// Study Plan : Day 50
// =============================================================
//
// QUESTION:
//   Partition string s such that every substring is a palindrome. Return all possible partitions.
// =============================================================

package main
import "fmt"
func partition(s string) [][]string {
    var res [][]string; cur := []string{}
    isPal := func(x string) bool { for i,j:=0,len(x)-1; i<j; i,j=i+1,j-1 { if x[i]!=x[j] { return false } }; return true }
    var bt func(int)
    bt = func(i int){
        if i == len(s) { tmp := make([]string, len(cur)); copy(tmp, cur); res = append(res, tmp); return }
        for j := i+1; j <= len(s); j++ {
            sub := s[i:j]; if isPal(sub) { cur = append(cur, sub); bt(j); cur = cur[:len(cur)-1] }
        }
    }
    bt(0); return res
}
func main(){ fmt.Println(partition("aab")) }

File: Java/Lesson100_PalindromePartitioning.java
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 100 -- Palindrome Partitioning
// Category   : Backtracking
// Difficulty : Medium
// Study Plan : Day 50
// =============================================================
//
// QUESTION:
//   Partition string s such that every substring is a palindrome. Return all possible partitions.
// =============================================================

import java.util.*;
public class Lesson100_PalindromePartitioning {
    List<List<String>> res = new ArrayList<>();
    boolean isPal(String s){ for (int i=0,j=s.length()-1;i<j;i++,j--) if (s.charAt(i)!=s.charAt(j)) return false; return true; }
    void bt(String s, int i, List<String> cur){
        if (i==s.length()){ res.add(new ArrayList<>(cur)); return; }
        for (int j=i+1;j<=s.length();j++){
            String sub = s.substring(i,j);
            if (isPal(sub)){ cur.add(sub); bt(s,j,cur); cur.remove(cur.size()-1); }
        }
    }
    public List<List<String>> partition(String s){ bt(s,0,new ArrayList<>()); return res; }
    public static void main(String[] a){ System.out.println(new Lesson100_PalindromePartitioning().partition("aab")); }
}
